consensus/tendermint/rule_proposal_and_polka_current_test.go

```diff
@@ -0,0 +1,98 @@
+package tendermint
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestProposalAndPolkaCurrent(t *testing.T) {
+	t.Run("Line 36: lock and broadcast precommit when step is prevote", func(t *testing.T) {
+		stateMachine := setupStateMachine(t, 4, 3)
+		currentRound := newTestRound(t, stateMachine, 0, 0)
+
+		committedValue := value(42)
+
+		// Initialise the round
+		currentRound.start()
+
+		// Prevotes are collected
+		currentRound.validator(0).prevote(&committedValue)
+		currentRound.validator(1).prevote(&committedValue)
+
+		// Receives proposal, go straight to precommit
+		currentRound.validator(0).proposal(committedValue, -1).expectActions(
+			currentRound.action().scheduleTimeout(prevote),
+			currentRound.action().broadcastPrevote(&committedValue),
+			currentRound.action().broadcastPrecommit(&committedValue),
+		)
+
+		assert.True(t, stateMachine.state.lockedValueAndOrValidValueSet)
+		assertState(t, stateMachine, height(0), round(0), precommit)
+	})
+
+	t.Run("Line 36: record valid value even if we don't prevote it", func(t *testing.T) {
+		stateMachine := setupStateMachine(t, 4, 1)
+		currentRound := newTestRound(t, stateMachine, 0, 0)
+		nextRound := newTestRound(t, stateMachine, 0, 1)
+		committedValue := value(42)
+
+		// Initialise the round
+		currentRound.start()
+
+		// Prevotes are collected but not proposal
+		currentRound.processTimeout(propose).expectActions(
+			currentRound.action().broadcastPrevote(nil),
+		)
+		currentRound.validator(2).prevote(&committedValue)
+		currentRound.validator(3).prevote(&committedValue)
+
+		currentRound.processTimeout(prevote).expectActions(
+			currentRound.action().broadcastPrecommit(nil),
+		)
+
+		// Receives proposal and prevote from validator 0, record valid value and round
+		currentRound.validator(0).proposal(committedValue, -1)
+		currentRound.validator(0).prevote(&committedValue)
+		assertState(t, stateMachine, height(0), round(0), precommit)
+		assert.Equal(t, &committedValue, stateMachine.state.validValue)
+		assert.Equal(t, round(0), stateMachine.state.validRound)
+
+		// Use valid value on next round
+		currentRound.validator(2).precommit(&committedValue)
+		currentRound.validator(3).precommit(&committedValue).expectActions(
+			currentRound.action().scheduleTimeout(precommit),
+		)
+		currentRound.processTimeout(precommit).expectActions(
+			nextRound.action().broadcastProposal(committedValue, 0),
+			nextRound.action().broadcastPrevote(&committedValue),
+		)
+	})
+
+	t.Run("Line 36: not trigger if not first time", func(t *testing.T) {
+		stateMachine := setupStateMachine(t, 4, 3)
+		currentRound := newTestRound(t, stateMachine, 0, 0)
+
+		committedValue := value(42)
+
+		// Initialise the round
+		currentRound.start()
+
+		// 2 prevotes are collected
+		currentRound.validator(0).proposal(committedValue, -1).expectActions(
+			currentRound.action().broadcastPrevote(&committedValue),
```

Comment:
What do you think about adding additional state assertions? Too Verbose? I feel like it makes it easier to check we are performing all the actions required by the tendermint algorithm, but it does add more code..

```suggestion
		currentRound.validator(0).proposal(committedValue, -1).expectActions( // line 22
			currentRound.action().broadcastPrevote(&committedValue),
		)
		assertState(t, stateMachine, height(0), round(0), prevote)
```